File: packages/python-sdk/src/services/agent_service.py

from decimal import Decimal
from typing import List
from web3 import Web3
from web3.contract import Contract
from eth_typing import Address
from ..types import AgentData

class AgentService:
    def __init__(self, agent_registry: Contract, account: Address):
        self.agent_registry = agent_registry
        self.account = account

    async def get_address(self) -> str:
        return self.account.address

    async def register_agent(self, model: str, prompt: str, skills: List[str]) -> str:
        tx = await self.agent_registry.functions.registerAgent(
            model, prompt, skills
        ).transact({'from': self.account})
        receipt = await self.agent_registry.web3.eth.wait_for_transaction_receipt(tx)
        
        event = self._find_event_in_receipt(receipt, "AgentRegistered")
        if not event or not event['args']:
            raise Exception("Agent registration failed")
        return event['args'][0]

    async def get_agent_data(self, agent_address: str) -> AgentData:
        data = await self.agent_registry.functions.getAgentData(agent_address).call()
        is_registered = await self.agent_registry.functions.isRegistered(agent_address).call()
        
        return AgentData(
            address=agent_address,
            model=data[0],
            prompt=data[1],
            skills=data[2],
            reputation=Decimal(data[3]),
            is_registered=is_registered
        )

    async def is_agent_registered(self, agent_address: str) -> bool:
        return await self.agent_registry.functions.isRegistered(agent_address).call()

    async def add_agent_skill(self, name: str, level: int) -> None:
        tx = await self.agent_registry.functions.addSkill(name, level).transact({'from': self.account})
        await self.agent_registry.web3.eth.wait_for_transaction_receipt(tx)

    async def update_agent_reputation(self, reputation: Decimal) -> None:
        tx = await self.agent_registry.functions.updateReputation(int(reputation)).transact({'from': self.account})
        await self.agent_registry.web3.eth.wait_for_transaction_receipt(tx)

    def _find_event_in_receipt(self, receipt: dict, event_name: str) -> Optional[dict]:
        for log in receipt['logs']:
            try:
                event = self.agent_registry.events[event_name]().process_log(log)
                return event
            except:
                continue
        return None

File: packages/sdk/src/services/AgentService.ts
import { BigNumberish, ethers } from "ethers";
import { AgentData } from "../types";

export class AgentService {
  private agentRegistry: ethers.Contract;
  private signer: ethers.Signer;
  
  constructor(agentRegistry: ethers.Contract, signer: ethers.Signer) {
    this.agentRegistry = agentRegistry;
    this.signer = signer;
  }

  /**
   * Gets the address of the agent.
   * @returns {Promise<string>} A promise that resolves to the agent address.
   */
  async getAddress(): Promise<string> {
    return this.signer.getAddress();
  }

  /**
   * Registers a new agent.
   * @param {string} model - The model of the agent.
   * @param {string} prompt - The prompt for the agent.
   * @param {string[]} skills - The skills of the agent.
   * @returns {Promise<string>} A promise that resolves to the agent address.
   */
  async registerAgent(model: string, prompt: string, skills: string[]): Promise<string> {
    const tx = await this.agentRegistry.registerAgent(model, prompt, skills);
    const receipt = await tx.wait();
    
    const event = this.findEventInReceipt(receipt, "AgentRegistered");
    if (!event?.args) {
      throw new Error("Agent registration failed");
    }
    const agentAddress = event.args[0];
    return agentAddress;
  }

  findEventInReceipt(receipt: any, eventName: string): ethers.EventLog {
    const events = receipt.logs.map((log: any) => {
      try {
        const event = this.agentRegistry.interface.parseLog(log);
        return event;
      } catch (e) {
        console.error('error:', e);
        return null;
      }
    }).filter((event: any) => event !== null);
    const event = events?.find((e: { name: string }) => e.name === eventName);
    return event;
  }

  /**
   * Gets data for a specific agent.
   * @param {string} agentAddress - The address of the agent.
   * @returns {Promise<AgentData>} A promise that resolves to the agent data.
   */
  async getAgentData(agentAddress: string): Promise<AgentData> {
    const [model, prompt, skills, reputation] = await this.agentRegistry.getAgentData(agentAddress);
    const isRegistered = await this.agentRegistry.isRegistered(agentAddress);

    return {
      address: agentAddress,
      model,
      prompt,
      skills,
      reputation,
      isRegistered
    };
  }

  /**
   * Checks if an agent is registered.
   * @param {string} agentAddress - The address of the agent.
   * @returns {Promise<boolean>} A promise that resolves to a boolean indicating if the agent is registered.
   */
  async isAgentRegistered(agentAddress: string): Promise<boolean> {
    return await this.agentRegistry.isRegistered(agentAddress);
  }

  /**
   * Adds a skill to an agent.
   * @param {string} name - The name of the skill.
   * @param {number} level - The level of the skill.
   * @returns {Promise<void>} A promise that resolves when the skill is added.
   */
  async addAgentSkill(name: string, level: number): Promise<void> {
    const tx = await this.agentRegistry.addSkill(name, level);
    await tx.wait();
  }


    /**
   * Updates the reputation of an agent.
   * @param {BigNumberish} reputation - The new reputation value.
   * @returns {Promise<void>} A promise that resolves when the reputation is updated.
   */
  async updateAgentReputation(reputation: BigNumberish): Promise<void> {
    const tx = await this.agentRegistry.updateReputation(reputation);
    await tx.wait();
  }
} 